### PR TITLE
Support used self-signed certificates LDAP.

### PR DESCRIPTION
**Pull request by veily**
support used self-signed certificates LDAP. #1277

### CODE REVIEW

File: connector/ldap/ldap.go [hunk 1]
```diff
@@ -68,7 +68,10 @@ type Config struct {
 
 	// Path to a trusted root certificate file.
 	RootCA string `json:"rootCA"`
-
+	// Path to a self-signed certificates ca file.
+	ClientCA string `json:"clientCA"`
```
Comment (srenatus) [inline connector/ldap/ldap.go]: ❓ Wouldn't this mostly match the `RootCA`? I'm not sure I understand the usecase where it differs. Can you please elaborate?

File: connector/ldap/ldap.go [hunk 2]
```diff
@@ -68,7 +68,10 @@ type Config struct {
...
+	//Path to a self-signed certificates private key file.
```
Comment (srenatus) [inline connector/ldap/ldap.go]: [nit] Can we make that `// Path` (with a space)?

**Comment by srenatus** [comment]
@veily Thanks for your contribution 😃

File: connector/ldap/ldap.go (hunk 1 above)
Comment (veily) [reply connector/ldap/ldap.go]: If ldap server setting verify client cert is true, only setting RootCA cannot work.

File: connector/ldap/ldap.go [hunk 3]
```diff
@@ -68,7 +68,10 @@ type Config struct {
 
 	// Path to a trusted root certificate file.
 	RootCA string `json:"rootCA"`
-
+	// Path to a self-signed certificates cert file.
```
Comment (srenatus) [inline connector/ldap/ldap.go]: Can we remove the "self-signed" there? I don't think it is necessary to have self-signed cert here.

Comment (veily) [reply connector/ldap/ldap.go]: Sorry, I am not clear enough. The self-signed certificate is not dex. The ldap server uses a self-signed certificate (RootCA, server cert) and opens the certificate for the client. In this case, if dex Ability to authenticate client certificates through the ldap server, using client certificates generated by the Ldap server RootCA

Comment (veily) [reply connector/ldap/ldap.go]: http://www.openldap.org/doc/admin24/tls.html
> 16.1.2. Client Certificates
The DN of a client certificate can be used directly as an authentication DN.
...

Comment (veily) [reply connector/ldap/ldap.go]: without the client certificates, dex all requests to ldap server will be rejected.

**Comment by veily** [comment]
Used for LDAP server that used self-signed certificates (RootCA, server cert) and setting TLS verify client. In this case, dex should use the client certificates generated by the Ldap server RootCA.
Without the client certificates, dex all requests to ldap server will be rejected.


http://www.openldap.org/doc/admin24/tls.html
> 16.1.2. Client Certificates
The DN of a client certificate can be used directly as an authentication DN.
...

File: connector/ldap/ldap.go [hunk 4]
```diff
@@ -225,6 +227,14 @@ func (c *Config) openConnector(logger logrus.FieldLogger) (*ldapConnector, error
 		}
 		tlsConfig.RootCAs = rootCAs
 	}
+
+	if c.ClientKey != "" && c.ClientCert != "" {
+		cert, err := tls.LoadX509KeyPair(c.ClientCert, c.ClientKey)
+		if err != nil {
+			return nil, fmt.Errorf("ldap: load self signed certs file failed.error:%s", err)
```
Comment (srenatus) [inline connector/ldap/ldap.go]: [nit] Can we please make that 
```go
fmt.Errorf("ldap: load client cert failed: %v", err)
```
just to be in line with the updated comment (thanks, btw), and the other error printouts above?

**Comment by ericchiang** [comment]
FYI there are commits that aren't attributed to the author of this PR. Please squash.

**Comment by srenatus** [comment]
@veily If you need help squashing commits, let me know! Once squashed, and having only you as author, we'll get this in 🎉

**Comment by veily** [comment]
@srenatus Okay, Please squash.  In addition, I used different git username  during  the completion of this PR, this is my mistake.

**Comment by srenatus** [comment]
Ah, I see. Could you please squash and do a force push onto your branch
(the one from this pr)? We could then merge it using the normal "merge"
button on github. Just to be in line with how we normally merge things.

**Comment by veily** [comment]
@srenatus Thanks, I has squashed commits under your guidance.

**Comment by srenatus** [comment]
Thank you! :)

**Pull request merged**